From Hollis Blanchard:

[] There is a theoretical bug fixed in the attached patch.
[] hvsi_drain_input() was calling yield() under a spinlock.

Index: linux-2.6.5/drivers/char/hvsi.c
===================================================================
--- linux-2.6.5.orig/drivers/char/hvsi.c
+++ linux-2.6.5/drivers/char/hvsi.c
@@ -898,14 +898,14 @@ static void hvsi_close(struct tty_struct
 		/* hvsi_write_worker will re-schedule itself until outbuf is empty. */
 		hvsi_flush_output(hp);
 
-		spin_lock_irqsave(&hp->lock, flags);
-
 		/* tell FSP to stop sending data */
 		hvsi_close_protocol(hp);
 
 		/* drain anything FSP is still in the middle of sending */
 		hvsi_drain_input(hp);
 
+		spin_lock_irqsave(&hp->lock, flags);
+
 		hp->tty = NULL;
 		hp->inbuf_end = hp->inbuf; /* discard remaining partial packets */
 	} else if (hp->count < 0)
